File: Temporada02/Episodio04/Ruby/Ejemplo/EjemploAnimal.rb

class Animal
  def self.hola
    'Hola! Soy un ' + self.tipo_de_animal
  end

  def self.tipo_de_animal
    'Implementar en las subclasses!'
  end
end

class Perro < Animal
  def self.hola
    super + ' gua gua'
  end
  def self.tipo_de_animal
    'Perro'
  end
end

class Gato < Animal
  def self.tipo_de_animal
    'Gato'
  end
end

class Veterinaria
  def hola_desde(un_tipo_de_animal)
    un_tipo_de_animal.hola
  end
end

puts Perro.tipo_de_animal
puts Gato.tipo_de_animal

una_veterinaria = Veterinaria.new
puts una_veterinaria.hola_desde Perro
puts una_veterinaria.hola_desde Gato

## IMPLEMENTAR hola en Perro ##

File: Temporada02/Episodio04/JavaScript/Ejemplo/EjemploAnimal.js

class Animal {
    static hola(){
        return 'Hola! Soy un ' + this.tipoDeAnimal()
    }

    static tipoDeAnimal(){
        throw new Error('Implementar en subclasse')
    }
}

class Perro extends Animal {
    static tipoDeAnimal() {
        return 'Perro'
    }
}

class Gato extends Animal {
    static tipoDeAnimal() {
        return 'Gato'
    }
}

class Veterinaria {
    holaDesde(unTipoDeAnimal) {
        return unTipoDeAnimal.hola()
    }
}


console.log(Perro.tipoDeAnimal())
console.log(Gato.tipoDeAnimal())

const unaVeterinaria = new Veterinaria()
console.log(unaVeterinaria.holaDesde(Perro))
console.log(unaVeterinaria.holaDesde(Gato))

// IMPLEMENTAR hola en Perro
